### PR TITLE
Add workflow to label community PRs and Issues.

### LINKED ISSUE #34197: Automatically Label Community Issues & Pull Requests
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**Issue Description:**

In order to better identify community contributions, we should automatically label issues and pull requests that have not originated from an account in one of Automattic's teams. There are likely some existing GitHub Actions we can utilize. I'm not sure what label would fit best, maybe another `type` label? `type: community contribution`?

### PR DESCRIPTION
**Pull request by jonathansadowski**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR creates a new GitHub workflow that labels community PRs and Issues.

Closes #34197 .

### How to test the changes in this Pull Request:

1. Fork this repo. All the steps here on out will be done in your fork ONLY!
1. Enable Actions ( by default it is disabled ). `settings/actions`.
1. Optional: you may want to disable some of the non-related actions, as they'll slow down execution of this action.
1. Enable Issues `settings`. 
1. In `trunk` of your fork, merge this branch `fix/34197`.
1. Create an issue in your fork authored by a user in either the `automattic` or `woocommerce` organization. Verify that the action runs, but that it skips labeling the issue because of the user being in either organization.
1. Create a PR in your fork authored by a user in either the `automattic` or `woocommerce` organization. Verify that the action runs, but that it skips labeling the PR because of the user being in either organization.
1. Create an issue in your fork authored by a user in neither the `automattic` nor the `woocommerce` organization. Verify that the action runs and that the issue is labeled as a community issue because of the user being in neither organization.
1. Create a PR in your fork authored by a user in neither the `automattic` nor the `woocommerce` organization. Verify that the action runs and that the PR is labeled as a community issue because of the user being in neither organization.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: .github/workflows/community-label.yml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    steps:
+      - name: label
+        uses: actions/github-script@v6
+        with:
+          script: |
+            github.rest.issues.addLabels( {
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              issue_number: context.issue.number,
+              labels: [ 'type: community contribution' ]
```
Comment (jonathansadowski) [inline .github/workflows/community-label.yml:59]: Note: We need to decide on this label before merging.

File: .github/workflows/community-label.yml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      issueId: ${{ steps.check.outputs.issueId }}
+      run: ${{ steps.check.outputs.run }}
+    steps:
+      - name: Check if user is a community contributor
+        id: check
+        uses: actions/github-script@v6
+        with:
+          script: |
+            const types = [ 'issue', 'pull_request' ];
+            const orgs = [ 'automattic', 'woocommerce' ];
```
Comment (jonathansadowski) [inline .github/workflows/community-label.yml]: Note: this currently doesn't exclude issues created by GlobalStep, as they do not have an organization for their users.

File: .github/workflows/community-label.yml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    outputs:
+      issueId: ${{ steps.check.outputs.issueId }}
+      run: ${{ steps.check.outputs.run }}
+    steps:
+      - name: Check if user is a community contributor
+        id: check
+        uses: actions/github-script@v6
+        with:
+          script: |
+            const types = [ 'issue', 'pull_request' ];
```
Comment (samueljseay) [inline .github/workflows/community-label.yml]: This script is fairly complex, any reason why we don't put this in its own file?

It bothers me that we don't get syntax highlighting and debugging/maintaining will be hard with the script being inline like this.

It looks like you can implement that pattern quite easily with this action: https://github.com/actions/github-script#run-a-separate-file

File: .github/workflows/community-label.yml [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            const orgs = [ 'automattic', 'woocommerce' ];
+
+            let found = null; // null indicates we haven't done any checks.
+            let type, org, result;
+
+            for ( var typeIndex in types ) {
```
Comment (samueljseay) [inline .github/workflows/community-label.yml]: We should use `const` everywhere possible. And if for some reason you don't want a const you should always use `let` as it respects block scoping.

File: .github/workflows/community-label.yml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+              type = types[ typeIndex ];
```
Comment (samueljseay) [inline .github/workflows/community-label.yml]: You've jumped through a lot of hoops here just to get the current array item of `types`. How about using [for ... of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) instead and removing this variable?

It should be:

```javascript
for ( const type of types ) {
  // access type here
}
```

File: .github/workflows/community-label.yml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                  try {
+                    result = await github.request( 'GET /orgs/{org}/public_members/{username}', {
+                      org,
+                      username: context.payload[ type ].user.login,
+                    } );
+                  } catch ( error ) {
+                    if ( error.status === 404 ) {
+                      // This means the user wasn't found in the organization.
+                      found = false;
+                      continue;
```
Comment (samueljseay) [inline .github/workflows/community-label.yml:74]: many breaks and continues are really not idiomatic JS either, I think this is a symptom of the fact you're trying to maintain state in let variables instead of using JavaScripts many collection filtering/manipulation options available. In my opinion it is very hard to follow when doing follow up maintenance work on this kind of code.

File: .github/workflows/community-label.yml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+              if ( ! context.payload[ type ] || ! context.payload[ type ].user || ! context.payload[ type ].user.login ) {
+                // If we don't have the context for this type, move on to checking the next type.
+                console.log( 'No payload found for %s, continuing...', type );
+                continue;
+              }
+
+              if ( context.payload[ type ].user.type && context.payload[ type ].user.type === 'Bot' ) {
+                // Treat bots as non-community contributors.
+                console.log( 'User %s is a bot user. Skipping...', payload[ type ].user.login );
+                found = true;
```
Comment (samueljseay) [inline .github/workflows/community-label.yml]: This is not really the idiomatic way to do this kind of work in JavaScript. I would typically avoid mutating variables in a loop generally and I would also avoid lots of break; and continue; But I think at least the `found` variable is ok here.

I can certainly see there is some complexity here though, you're nesting loops and using async code.

I think I would keep the `let found` in this case and break out the async/sync code into functions something like:

```
async function userIsInAnOrgWeCareAbout(username) {
     for (const org of orgs) {
      ....
     } 
}

function userIsBot(contextPayload, type) {
}

function validPayload(contextPayload, type) {
  // If we don't have the context for this type, move on to checking the next type.
}
```

And then the code should be something like:

```
let found = false;
for (const type of types) {
  // rather than using break and continue just don't do anything if payload is not there
  if(validPayload(context.payload)) {
    if (userIsBot(context.payload, type)) {
      found = true;
    } else {
       found = await userIsInOrgWeCareAbout(context.payload[ type ].user.login);
    }
  }
}
```

**Review by samueljseay — CHANGES_REQUESTED** [review]
I think the script code needs some clean up and to be a bit more idiomatic.

Also in testing, I found that it did not correctly skip labelling for my user (in both Automattic and Woo orgs). I *think* I did everything correctly.

See https://github.com/samueljseay/woocommerce/issues/3

File: .github/workflows/community-label.yml (hunk 3 above)
Comment (jonathansadowski) [reply .github/workflows/community-label.yml]: Primarily because of the checkout requirement and execution time, but I'll give that a try and keep an eye on performance.

**Comment by jonathansadowski** [comment]
Thanks for the review @samueljseay. I pushed updates based on your suggestion.

It should also be possible to test the module introduced in this update something like this:

```
const isCommunityContributor = require( './.github/workflows/scripts/is-community-contributor' );
const { Octokit } = require( '@octokit/rest' );
const github = new Octokit( {
	auth: 'YOURTOKEN',
} );

const context = {
	payload: {
		issue: {
			user: {
				login: 'jonathansadowski',
			}
		}
	}
};

const config = {
	orgs: [ 'woocommerce', 'automattic' ],
	types: [ 'pull_request', 'issue' ],
};

( async ( { github, context, config } ) => {
	try {
		const isCommunity = await isCommunityContributor( { github, context, config } );
		console.log( '::set-output name=run::%s', isCommunity ? 'true' : 'false' );
	} catch ( e ) {
		console.log( '::set-output name=run::false' );
	}
} )( { github, context, config } );
```

Where you'd replace `YOURTOKEN` with a Github token that has `repo` and `read:org` scopes.

We may need to verify the scopes available to the automatically generated token before merging this. It's possible we need to use a token with `read:org` scope if that's not available to the generated token. That might have been why you ran into issues in your first pass at testing this, as neither woocommerce nor automattic are public memberships for you.

**Review by samueljseay — DISMISSED** [review]
Code looks good! and thanks for the test script, that saved me a lot of time! I tested my user, your user and a user I know is not in the orgs and it worked as expected 🚢

**Comment by samueljseay** [comment]
You may just want to merge trunk to your branch to get the checks passing.
